xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-default/src/main/java/org/xwiki/mentions/internal/DefaultMentionsDataConsumer.java

```diff
@@ -0,0 +1,300 @@
+import static java.util.Collections.emptyList;
+import static java.util.Optional.ofNullable;
+import static java.util.stream.Collectors.toList;
+import static org.apache.commons.lang3.exception.ExceptionUtils.getRootCauseMessage;
+import static org.xwiki.annotation.Annotation.SELECTION_FIELD;
+import static org.xwiki.mentions.MentionLocation.ANNOTATION;
+import static org.xwiki.mentions.MentionLocation.AWM_FIELD;
+import static org.xwiki.mentions.MentionLocation.COMMENT;
+
+/**
+ * Default implementation of {@link MentionsDataConsumer}.
+ *
+ * @version $Id$
+ * @since 12.6RC1
+ */
+@Component
+@Singleton
+public class DefaultMentionsDataConsumer implements MentionsDataConsumer
+{
+    @Inject
+    private DocumentReferenceResolver<String> documentReferenceResolver;
+
+    @Inject
+    private ExecutionContextManager contextManager;
+
+    @Inject
+    private Provider<XWikiContext> xcontextProvider;
+
+    @Inject
+    private Execution execution;
+
+    @Inject
+    private DocumentRevisionProvider documentRevisionProvider;
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private MentionNotificationService notificationService;
+
+    @Inject
+    private MentionXDOMService xdomService;
+
+    /**
+     * Initialize the context.
+     *
+     * @param authorReference the author of the analyzed document
+     * @param wikiId the wiki id
+     * @throws ExecutionContextException in case one {@link ExecutionContextInitializer} fails to execute 
+     *
+     */
+    private void initContext(DocumentReference authorReference, String wikiId) throws ExecutionContextException
+    {
+        ExecutionContext context = new ExecutionContext();
+        this.contextManager.initialize(context);
+
+        XWikiContext xWikiContext = this.xcontextProvider.get();
+        xWikiContext.setUserReference(authorReference);
+        xWikiContext.setWikiReference(authorReference.getWikiReference());
+        xWikiContext.setWikiId(wikiId);
+    }
+
+    @Override
+    public void consume(MentionsData data) throws XWikiException
+    {
+        try {
+            DocumentReference author = this.documentReferenceResolver.resolve(data.getAuthorReference());
+            this.initContext(author, data.getWikiId());
+            DocumentReference dr = this.documentReferenceResolver.resolve(data.getDocumentReference());
+            XWikiDocument doc =
+                this.documentRevisionProvider.getRevision(dr, data.getVersion());
+            XWikiDocument oldDoc;
+            if (doc.getPreviousVersion() != null) {
+                oldDoc = this.documentRevisionProvider
+                             .getRevision(dr, doc.getPreviousVersion());
+            } else {
+                oldDoc = null;
+            }
+            DocumentReference documentReference = doc.getDocumentReference();
+            DocumentReference authorReference = doc.getAuthorReference();
+
+            String documentReferenceStr = documentReference.toString();
```

Comment:
So that's not good, you should use a call to a serializer. Now I still think that it should be serialized in sendNotif and you should continue to propagate DocumentReference type.